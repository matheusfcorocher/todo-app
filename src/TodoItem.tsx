import React from "react";

type TodoItemStates = "COMPLETED" | "ACTIVE" | "EDITING";

interface TodoItemProps {
    task: {
        id: string,
        title: string,
        state: TodoItemStates
    },
    onCompletingTodo: () => void;
    onEditingTodo: () => void;
}

export const TodoItem = ({
    task: {
        id,
        title,
        state
    },
    onCompletingTodo,
    onEditingTodo
} : TodoItemProps) => {
    return(
        <li className="todo-item">
            <label htmlFor="title" aria-label={title}>
                <input type="text" value={title} name="title"/>
            </label>
        </li>
    );
}
