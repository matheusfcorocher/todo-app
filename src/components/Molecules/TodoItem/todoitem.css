.todo-item {
    display: flex;
    justify-content: space-between;
    width: auto;
    max-height: 3rem;
    padding: 0.5rem;
    box-shadow: #00000026 0px 0px 0px 1px inset;
    background: #826a72;
}

form {
    display: flex;
    margin: auto;
    align-items: center;
    justify-content: flex-start;
    width: 100%;
    height: auto;
}

.EDITING .checkbox {
    visibility: hidden;
}

.delete-button {
    visibility: hidden;
}

.todo-item:hover .delete-button {
    visibility: visible;
}
