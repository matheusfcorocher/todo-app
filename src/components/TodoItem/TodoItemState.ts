import React, { useState } from "react";

const [todoItemTitle, setTodoItemTitle] = useState(title);
const [todoItemPastState, setTodoItemPastState] = useState(state);
const [todoItemState, setTodoItemState] = useState(state);

const updateTaskState = (newState: string) => {
    setTodoItemState(newState);
};

const updateTaskTitle = () => {
    return {
        onChange: updateTaskTitleOnChange,
        onBlur: updateTaskTitleOnBlur,
        onFocus: updateTaskTitleOnFocus,
    }
}

const updateTaskTitleOnChange = (newTitle: string) => {
    setTodoItemTitle(newTitle);
};

const updateTaskTitleOnBlur = () => {
    setTodoItemState(todoItemPastState == "EDITING" ? "ACTIVE" 
    : todoItemPastState);
}

const updateTaskTitleOnFocus = () => {
    setTodoItemPastState(todoItemState);
    setTodoItemState("EDITING");
}