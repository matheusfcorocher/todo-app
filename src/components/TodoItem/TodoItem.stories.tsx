import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import { TodoItem } from './TodoItem';

export default {
  title: 'Example/TodoItem',
  component: TodoItem,
  decorators: [
    (Story) => (
      <ul style={{ listStyleType: 'none', margin: 0, padding: 0}}>
        <Story />
      </ul>
    ),
  ]
} as ComponentMeta<typeof TodoItem>;

const Template: ComponentStory<typeof TodoItem> = (args) => <TodoItem {...args} />;

export const Default = Template.bind({});
Default.args = {
    task: {
        id: "1",
        title: "TASK",
        state: "ACTIVE",
    }
};

export const Editing = Template.bind({});
Editing.args = {
    task: {
        id: "2",
        title: "TASK",
        state: "EDITING",
    }
};

export const Completed = Template.bind({});
Completed.args = {
    task: {
        id: "3",
        title: "TASK",
        state: "COMPLETED",
    }
};