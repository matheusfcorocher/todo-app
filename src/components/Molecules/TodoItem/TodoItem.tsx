import React, { useRef, useState, useEffect } from "react";
import deleteIcon from '../../../assets/icons/delete.svg';

import './todoitem.css';

interface TodoItemProps {
    task: {
        id: string,
        title: string,
        isCompleted: boolean
    },
    handleUpdateTodoItemTitle: (newTitle: string) => void,
    handleUpdateTodoItemState: (newState: boolean) => void,
    handleDeleteTodoItem: (newState: string) => void,
}

export const TodoItem = ({
    task: {
        id,
        title,
        isCompleted
    },
    handleUpdateTodoItemTitle,
    handleUpdateTodoItemState,
    handleDeleteTodoItem,
}: TodoItemProps) => {
    const ref = useRef() as React.MutableRefObject<HTMLInputElement>;
    const [hasFocus, setFocus] = useState(false);

    useEffect(() => {
        if (document.hasFocus() && ref.current.contains(document.activeElement)) {
            setFocus(true);
        }
    }, []);

    return (
        <li 
            data-testid={`todoItem-${id}`}
            className={`todo-item ${isCompleted}`}
        >
            <form>
                <label
                    htmlFor="checked"
                    aria-label={`todoItem-${id}`}
                    className={`checkbox`}
                >
                    <input
                        type="checkbox"
                        role="checkbox"
                        name="checked"
                        id={`todoItem-${id}`}
                        className="checkbox-input"
                        checked={isCompleted}
                        onChange={() => {
                            if(isCompleted) {
                                handleUpdateTodoItemState(false);
                            } else {
                                handleUpdateTodoItemState(true);
                            }
                        }}
                    />
                </label>
                <label
                    htmlFor="title"
                    aria-label={title}
                    className="title"
                >
                    <input
                        type="text"
                        role="textbox"
                        value={title}
                        className={`title-input`}
                        ref={ref}
                        onChange={() => handleUpdateTodoItemTitle}
                        onFocus={() => setFocus(true)}
                        onBlur={() => setFocus(false)}
                    />
                </label>
                {!hasFocus && (
                    <button
                        className="delete-button"
                        onClick={() => handleDeleteTodoItem}
                        id={`deleteTodoItem-${id}`}
                        role="button"
                        key={`deleteTodoItem-${id}`}
                    >
                        <img src={deleteIcon} className="delete-icon" alt="delete" />
                    </button>
                )}
            </form>
        </li>
    );
}
