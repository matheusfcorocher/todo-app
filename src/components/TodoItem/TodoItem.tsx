import React, { useState } from "react";
import deleteIcon from '../../assets/icons/delete.svg';

import './todoitem.css';

type TodoItemStates = "COMPLETED" | "ACTIVE" | "EDITING";

interface TodoItemProps {
    task: {
        id: string,
        title: string,
        state: TodoItemStates
    },
    handleUpdateTodoItemTitle: (newTitle: string) => void,
    handleUpdateTodoItemState: (newState: string) => void,
    handleDeleteTodoItem: (newState: string) => void,
}

export const TodoItem = ({
    task: {
        id,
        title,
        state
    },
    handleUpdateTodoItemTitle,
    handleUpdateTodoItemState,
    handleDeleteTodoItem,
}: TodoItemProps) => {
    return (
        <li className={`todo-item ${state}`}>
            <form>
                <label
                    htmlFor="checked"
                    aria-label={`todoItem-${id}`}
                    className={`checkbox`}
                >
                    <input
                        type="checkbox"
                        name="checked"
                        id={`todoItem-${id}`}
                        className="checkbox-input"
                        checked={state === "COMPLETED"}
                        onChange={() => handleUpdateTodoItemState}
                    />
                </label>
                <label
                    htmlFor="title"
                    aria-label={title}
                    className="title"
                >
                    <input
                        type="text"
                        value={title}
                        name="title"
                        className={`title-input`}
                        onChange={() => handleUpdateTodoItemTitle}
                        onFocus={() => handleUpdateTodoItemTitle}
                        onBlur={() => handleUpdateTodoItemTitle}
                    />
                </label>
                {state !== "EDITING" && (
                    <button
                        className="delete-button"
                        onClick={() => handleDeleteTodoItem}
                        id={`deleteTodoItem-${id}`}
                        aria-label={`deleteTodoItem-${id}`}
                        key={`deleteTodoItem-${id}`}
                    >
                        <img src={deleteIcon} className="delete-icon" alt="delete" />
                    </button>
                )}
            </form>
        </li>
    );
}
