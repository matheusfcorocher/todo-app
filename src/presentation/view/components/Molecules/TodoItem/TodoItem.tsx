import React, { useRef, useState, useEffect } from "react";
import { TodoTask } from "../../../../../domain/entities/TodoTask";
import { TodoTaskControllerReturnType } from "../../../../controllers/TodoTaskController";
import Checkbox from "../../Atoms/Checkbox/Checkbox";
import IconButton from "../../Atoms/IconButton/IconButton";
import DeleteIcon from "../../Atoms/icons/DeleteIcon/DeleteIcon";
import TitleInput from "../../Atoms/TitleInput/TitleInput";

import './todoitem.css';

interface TodoItemProps {
    task: TodoTask,
    todoTaskController: TodoTaskControllerReturnType;
}

export const TodoItem = ({
    task: {
        id,
        title,
        isCompleted
    },
    todoTaskController
}: TodoItemProps) => {
    const { handleUpdateTodoTaskState, handleUpdateTodoTaskTitle, handleDeleteTodoTask } = todoTaskController
    const ref = useRef<HTMLInputElement>(document.createElement('input'));
    const [hasFocus, setFocus] = useState(false);
    useEffect(() => {
        if (document.hasFocus() && ref.current.contains(document.activeElement)) {
            setFocus(true);
        }
    }, []);

    return (
        <li
            data-testid={`todoItem-${id}`}
            className={`todo-item ${isCompleted}`}
        >
            <form onSubmit={event => event.preventDefault()}>
                <Checkbox
                    handleOnChange={() => {
                        if (isCompleted) {
                            handleUpdateTodoTaskState(id, false);
                        } else {
                            handleUpdateTodoTaskState(id, true);
                        }
                    }}
                    checked={isCompleted}
                />
                <TitleInput
                    id={id}
                    title={title}
                    handleOnChange={handleUpdateTodoTaskTitle}
                    handleOnFocus={() => setFocus(true)}
                    handleOnBlur={() => setFocus(false)}
                    ref={ref}
                />
                {!hasFocus && (
                    <IconButton
                        className="delete-button"
                        handleFunction={() => handleDeleteTodoTask(id)}>
                        <DeleteIcon />
                    </IconButton>
                )}
            </form>
        </li>
    );
}
