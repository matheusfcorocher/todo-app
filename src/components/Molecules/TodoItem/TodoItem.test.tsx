import React, { useState } from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import { TodoItem } from './TodoItem';
import userEvent from '@testing-library/user-event';

describe("::Components ::TodoItem", () => {

  afterEach(() => {
    cleanup();
  })

  describe("When TodoItem isn't completed", () => {
    beforeEach(() => {
      render(
        <TodoItem task={{
          id: "1",
          title: "wash dishes",
          isCompleted: false,
        }}
        handleUpdateTodoItemTitle={function (newTitle: string): void {
          throw new Error('Function not implemented.');
        } }
        handleDeleteTodoItem={function (newState: string): void {
          throw new Error('Function not implemented.');
        } } handleUpdateTodoItemState={function (id: string, state: boolean): void {
          throw new Error('Function not implemented.');
        } } />);
    })
    test('renders checkbox unchecked', () => {
      const checkbox = screen.getByRole("checkbox");
      expect(checkbox).toBeInTheDocument();
      expect(checkbox).not.toBeChecked();
    })
    test('renders title of TodoItem', () => {
      const title = screen.getByRole("textbox");
      expect(title).toBeInTheDocument();
      expect(title).toHaveValue("wash dishes");
    })
    test('renders delete button when hover TodoItem', () => {
      const deleteButton = screen.getByRole(/button/i);
      const todoItem = screen.getByTestId(/todoItem-/i);
      userEvent.hover(todoItem);
      expect(deleteButton).toBeVisible();
    })
  });
  describe('When TodoItem has COMPLETE state', () => {
    beforeEach(() => {
      render(
        <TodoItem task={{
          id: "1",
          title: "wash dishes",
          isCompleted: true,
        }}
        handleUpdateTodoItemTitle={function (newTitle: string): void {
          throw new Error('Function not implemented.');
        } }
        handleDeleteTodoItem={function (newState: string): void {
          throw new Error('Function not implemented.');
        } } handleUpdateTodoItemState={function (id: string, state: boolean): void {
          throw new Error('Function not implemented.');
        } } />);
    })
    test('renders checkbox checked', () => {
      const checkbox = screen.getByRole("checkbox");
      expect(checkbox).toBeInTheDocument();
      expect(checkbox).toBeChecked();
    })
    test('renders title of TodoItem', () => {
      const title = screen.getByRole("textbox");
      expect(title).toBeInTheDocument();
      expect(title).toHaveValue("wash dishes");
    })
    test('renders delete button when hover TodoItem', () => {
      const deleteButton = screen.getByRole(/button/i);
      const todoItem = screen.getByTestId(/todoItem-/i);
      userEvent.hover(todoItem);
      expect(deleteButton).toBeVisible();
    })
  });
  describe('When TodoItem has EDITING state', () => {
    beforeEach(() => {
      render(
        <TodoItem task={{
          id: "1",
          title: "wash dishes",
          isCompleted: false,
        }}
        handleUpdateTodoItemTitle={function (newTitle: string): void {
          throw new Error('Function not implemented.');
        } }
        handleDeleteTodoItem={function (newState: string): void {
          throw new Error('Function not implemented.');
        } } handleUpdateTodoItemState={function (id: string, state: boolean): void {
          throw new Error('Function not implemented.');
        } } />);
    })
    test('delete button isnt showing', () => {
      const deleteButton = screen.getByRole("button", { name: "delete" });
      const input = screen.getByRole("textbox");
      userEvent.click(input);
      expect(deleteButton).not.toBeVisible();
    });
  });
  describe('When title of TodoItem is too long', () => {
    beforeEach(() => {
      render(
        <TodoItem task={{
          id: "1",
          title: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum accumsan suscipit turpis vel volutpat. In vitae aliquam ex. Nullam in odio pellentesque, feugiat libero vel, fringilla massa. Duis nec odio in nisl fringilla vestibulum at feugiat lorem. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas.",
          isCompleted: false,
        }}
        handleUpdateTodoItemTitle={function (newTitle: string): void {
          throw new Error('Function not implemented.');
        } }
        handleDeleteTodoItem={function (newState: string): void {
          throw new Error('Function not implemented.');
        } } handleUpdateTodoItemState={function (id: string, state: boolean): void {
          throw new Error('Function not implemented.');
        } } />);
    })
    test('renders title of TodoItem with ellipsis', () => {
      const input = screen.getByRole("textbox");
      expect(input).toMatchSnapshot();
    })
  });
})
