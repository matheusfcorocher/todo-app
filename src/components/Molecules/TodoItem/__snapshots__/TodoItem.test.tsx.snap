// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`::Components ::TodoItem When title of TodoItem is too long renders title of TodoItem with ellipsis 1`] = `
<input
  class="title-input"
  role="textbox"
  type="text"
  value="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum accumsan suscipit turpis vel volutpat. In vitae aliquam ex. Nullam in odio pellentesque, feugiat libero vel, fringilla massa. Duis nec odio in nisl fringilla vestibulum at feugiat lorem. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas."
/>
`;
